source/dub/logging.d

```diff
@@ -0,0 +1,384 @@
+		and whether to color outputted text or not can be set with
+		setLoggingColorsEnabled(..)
+
+		The color(str, color) function can be used to color text within a log
+		message, for instance like this:
+
+		logInfo("Tag", Color.green, "My %s message", "colored".color(Color.red))
+
+	Copyright: © 2018 Giacomo De Lazzari
+	License: Subject to the terms of the MIT license, as written in the included LICENSE file.
+	Authors: Giacomo De Lazzari
+*/
+
+module dub.logging;
+
+import std.stdio;
+import std.array;
+import std.format;
+import std.string;
+
+import dub.internal.colorize : fg, mode;
+
+/**
+	An enum listing possible colors for terminal output, useful to set the color
+	of a tag. Re-exported from d-colorize in dub.internal.colorize. See the enum
+	definition there for a list of possible values.
+*/
+public alias Color = fg;
+
+/**
+	An enum listing possible text "modes" for terminal output, useful to set the
+	text to bold, underline, blinking, etc...
+	Re-exported from d-colorize in dub.internal.colorize. See the enum definition
+	there for a list of possible values.
+*/
+public alias Mode = mode;
+
+/// The tag width in chars, defined as a constant here
+private const int TAG_WIDTH = 12;
+
+/// Possible log levels supported
+enum LogLevel {
+	debug_,
+	diagnostic,
+	info,
+	warn,
+	error,
+	none
+}
+
+// The current minimum log level to be printed
+private shared LogLevel _minLevel = LogLevel.info;
+
+/*
+	Whether to print text with colors or not, defaults to true but will be set
+	to false in initLogging() if stdout or stderr are not a TTY (which means the
+	output is probably being piped and we don't want ASCII escape chars in it)
+*/
+private shared bool _printColors = true;
+
+// isatty() is used in initLogging() to detect whether or not we are on a TTY
+extern (C) int isatty(int);
+
+/**
+	This function must be called at the beginning for the program, before any
+	logging occurs. It will detect whether or not stdout/stderr are a console/TTY
+	and will consequently disable colored output if needed.
+
+	Forgetting to call the function will result in ASCII escape sequences in the
+	piped output, probably an undesiderable thing.
+*/
+void initLogging()
+{
+	import core.stdc.stdio;
+
+	// Initially enable colors, we'll disable them during this functions if we
+	// find any reason to
+	_printColors = true;
+
+	// The following stuff depends on the platform
+	version (Windows)
+	{
+		version (CRuntime_DigitalMars)
+		{
+			if (!isatty(core.stdc.stdio.stdout._file) ||
+					!isatty(core.stdc.stdio.stderr._file))
+				_printColors = false;
+		}
+		else version (CRuntime_Microsoft)
+		{
+			if (!isatty(fileno(core.stdc.stdio.stdout)) ||
+					!isatty(fileno(core.stdc.stdio.stderr)))
+				_printColors = false;
+		}
+		else
+			_printColors = false;
+	}
+	else version (Posix)
+	{
+		import core.sys.posix.unistd;
+
+		if (!isatty(STDERR_FILENO) || !isatty(STDOUT_FILENO))
+			_printColors = false;
+	}
```

Comment:
I think it would be best to have a special `inTerminal` method, s.t. this can be simplified to: `setLoggingColorsEnabled(!inTerminal());`